snewpdag/plugins/Chi2Calculation.py

```diff
@@ -0,0 +1,204 @@
+"""
+Chi2Calculator: generates a skymap of SN direction probabilities
+
+Constructor Arguments:
+    detector_list: list of strings, ["first_detector", "second_detector", ...]    \
+                the list of detectors that we want to include in the calculations  \
+                options: "HK", "IC", "JUNO", "KM3", "SK"                           / same as in NeutrinoArrivalTime
+    detector_location: csv file name ('detector_location.csv')                  __/
+    output_file: name of file to save .fits map to (e.g. 'output/map.fits')
+    DET0: detector with the lowest error (delta_ts are calculated w.r.t. this detector) (e.g. 'SK')
+    NSIDE: (int) healpy map parameter, it describes map resolution (32 is a reasonable number)
+    add_contours: (bool) (optional argument, default is True) if True than 1-, 2- and 3-sigma countours
+                  are rescaled to make them visible (map is no longer a real chi^2 map), otherwise chi^2
+                  map without any rescaling is saved
+
+Output:
+    It doesn't add anything to data, saves a .fits file to <output_file>
+"""
+
+import csv
+import numpy as np
+import healpy as hp
+import math
+from scipy.stats import chi2
+from datetime import datetime
+
+from snewpdag.dag import Node
+
+
+class Chi2Calculator(Node):
+    def __init__(self, detector_list, detector_location,
+                 det0, NSIDE, output_file, **kwargs):
+        self.detector_info = {}
+        with open(detector_location, 'r') as f:
+            detectors = csv.reader(f)
+            for detector in detectors:
+                name = detector[0]
+                if name not in detector_list:
+                    continue
+                lon = np.radians(float(detector[1]))
+                lat = np.radians(float(detector[2]))
+                height = float(detector[3])
+                sigma = float(detector[4])
+                bias = float(detector[5])
+                if name == det0:
+                    self.d0_info = [
+                        lon, lat, height, sigma, bias, name]
+                else:
+                    self.detector_info[name] = [
+                        lon, lat, height, sigma, bias]
+        self.precision_matrix = self.generatePrecisionMatrix()
+        self.n_dof = len(detector_list) - 3
+        self.NSIDE = NSIDE
+        self.NPIX = hp.nside2npix(NSIDE)
+        self.output_file = output_file
+
+        if 'add_contours' in kwargs:
+            self.add_contours = kwargs.pop('add_contours')
+        else:
+            self.add_contours = True
+
+        super().__init__(**kwargs)
+
+    # Generates precision matrix (inverse of covariance matrix)
+    def generatePrecisionMatrix(self):
+        sigma_0 = self.d0_info[3]
+        n_detectors = len(self.detector_info)
+        V = [[sigma_0]*n_detectors for i in range(n_detectors)]
+        for i in range(n_detectors):
+            det = list(self.detector_info.keys())[i]
+            V[i][i] = self.detector_info[det][3]
```

Comment:
should this be squared (V is a matrix of variances, but sigma is a standard deviation)? And I think it should be added to the diagonal, not replacing it.  Similarly, I think the off-diagonal elements should be sigma_0^2, not sigma_0 itself.

There's a somewhat more numpy-flavored way of doing this in my document, page 3, mvar.